src/main/java/software/amazon/cloudformation/proxy/AmazonWebServicesClientProxy.java

```diff
@@ -174,23 +250,54 @@ public ClientT client() {
+                            ? super RequestT, Exception, ClientT, ModelT, CallbackT, Boolean> retryFilter) {
+                            return handleError(((request, exception, client_, model_, context_) -> {
+                                if (retryFilter.invoke(request, exception, client_, model_, context_)) {
+                                    throw RetryableException.builder().build();
                                 }
-                                return defaultHandler(request, exception, client1, model1, context1);
-                            });
+                                return defaultHandler(request, exception, client_, model_, context_);
+                            }));
                         }
 
                         @Override
-                        public Completed<RequestT, ResponseT, ClientT, ModelT, CallbackT> exceptHandler(Callback<? super RequestT,
-                            Exception, ClientT, ModelT, CallbackT, ProgressEvent<ModelT, CallbackT>> handler) {
-                            this.exceptHandler = handler;
+                        public Completed<RequestT, ResponseT, ClientT, ModelT, CallbackT> handleError(ExceptionPropagate<
+                            ? super RequestT, Exception, ClientT, ModelT, CallbackT, ProgressEvent<ModelT, CallbackT>> handler) {
+                            getExceptionHandler(handler);
                             return this;
                         }
 
+                        private
+                            Callback<? super RequestT, Exception, ClientT, ModelT, CallbackT, ProgressEvent<ModelT, CallbackT>>
+                            getExceptionHandler(final ExceptionPropagate<? super RequestT, Exception, ClientT, ModelT, CallbackT,
+                                ProgressEvent<ModelT, CallbackT>> handler) {
+                            if (this.exceptHandler == null) {
+                                this.exceptHandler = ((request, exception, client_, model_, context_) -> {
+                                    ProgressEvent<ModelT, CallbackT> event = null;
+                                    Exception ex = exception;
+                                    ExceptionPropagate<? super RequestT, Exception, ClientT, ModelT, CallbackT,
+                                        ProgressEvent<ModelT, CallbackT>> inner = handler;
+                                    boolean defaultHandler = false;
+                                    do {
+                                        try {
+                                            event = inner.invoke(request, ex, client_, model_, context_);
+                                        } catch (RetryableException e) {
+                                            break;
+                                        } catch (Exception e) {
+                                            if (defaultHandler) {
+                                                throw new TerminalException("FRAMEWORK ERROR, LOOPING cause " + e, e);
```

Comment:
Harsh error message, will that be surfaced to the caller?

Reply:
This is meant to be harsh, this is indicating a LOOP is happening, which should never happen. Open to suggestions about this